src/session/media_viewer.rs

```diff
@@ -0,0 +1,206 @@
+                        ((target_width - bounds.width() as f64) * perc) + bounds.width() as f64,
+                        ((target_height - bounds.height() as f64) * perc) + bounds.height() as f64,
+                    )
+                };
+
+                // Media
+                snapshot.save();
+                snapshot.translate(&graphene::Point::new(x as f32, y as f32));
+                let perc_rounding = ROUNDING - (ROUNDING * perc as f32);
+                let rounding_size = graphene::Size::new(perc_rounding, perc_rounding);
+                let rounding_bounds = gsk::RoundedRect::new(
+                    graphene::Rect::new(0.0, 0.0, width as f32, height as f32),
+                    rounding_size,
+                    rounding_size,
+                    rounding_size,
+                    rounding_size,
+                );
+                snapshot.push_rounded_clip(&rounding_bounds);
+                paintable.snapshot(snapshot.upcast_ref(), width, height);
+                snapshot.pop();
+                snapshot.restore();
+
+                // UI Overlay
+                snapshot.push_opacity(perc);
+                widget.snapshot_child(&*self.header_bar, snapshot);
+                snapshot.pop();
+            }
+        }
+    }
+}
+
+glib::wrapper! {
+    pub(crate) struct MediaViewer(ObjectSubclass<imp::MediaViewer>)
+        @extends gtk::Widget;
+}
+
+impl MediaViewer {
+    pub(crate) fn open_media(
+        &self,
+        paintable: gdk::Paintable,
+        target_widget: impl IsA<gtk::Widget>,
+    ) {
+        let imp = self.imp();
+        imp.paintable.replace(Some(paintable));
+
+        let bounds = target_widget.compute_bounds(self).unwrap();
+        imp.widget_bounds.replace(Some(bounds));
+
+        // Hide the widget so that we don't see the original media while transitioning.
+        // We don't use the visible property here because otherwise the widgets doesn't
+        // get allocated anymore.
+        target_widget.set_opacity(0.0);
+
+        imp.target_widget.replace(Some(target_widget.upcast()));
+
+        let animation = adw::TimedAnimation::builder()
+            .widget(self)
+            .value_from(0.0)
+            .value_to(1.0)
+            .duration(300)
+            .target(&adw::CallbackAnimationTarget::new(Some(Box::new(
+                clone!(@weak self as obj => move |_| {
+                    obj.queue_draw();
+                }),
+            ))))
+            .easing(adw::Easing::EaseOutQuart)
+            .build();
+        imp.animation.replace(Some(animation.clone()));
+
+        // Make the overlay visible first, so that the animation isn't automatically skipped
+        self.set_visible(true);
+
+        animation.play();
+    }
+
+    fn go_back(&self) {
+        let animation_ref = self.imp().animation.borrow();
+        let animation = animation_ref.as_ref().unwrap();
+
+        // Reverse the animation and go back to the previous state
```

Comment:
It's a good idea to update the target widget bounds, because meanwhile the message may have moved (e.g. by receiving new messages).